### ISSUE TITLE
Add sliding window attention to sdpa in mistral

### ISSUE THREAD

**Issue by ehuaa** [huggingface/transformers#28980]
### Feature request

https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L1006-L1023
![image](https://github.com/huggingface/transformers/assets/5137359/9601a5d2-cf9f-4ef6-a0ab-047a8cd7f1cd)

In the code listed above, the latest version of transformers cannot use sliding window feature in mistral model.
I doubt that the reason is you mentioned above,
https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L687-L688
![image](https://github.com/huggingface/transformers/assets/5137359/997cd770-e17f-4eb4-997b-fe65f30ddc85)
And this issue in PyTorch makes you bugged with custom attn_mask like sliding window attention mask.
https://github.com/pytorch/pytorch/issues/112577

While this issue has been fixed since torch 2.2.0, and it has been released two weeks ago, can you add this feature back to sdpa kernel in mistral?

### Motivation

I cannot use sliding window with sdpa right now, cause my gpu card is V100, i cannot work with flashattention2.

### Your contribution

I think we can pass sliding_window param to _prepare_4d_causal_attention_mask_for_sdpa function.

**Comment by amyeroberts** [comment]
cc @fxmarty

**Comment by fxmarty** [comment]
Hi, thank you for the suggestion, SDPA support for mistral was added by @ArthurZucker in https://github.com/huggingface/transformers/pull/28133.

**Comment by ArthurZucker** [comment]
I think it comes down to just adding `sliding_window` to the call for `_prepare_4d_causal_attention_mask_for_sdpa` yes. Would you like to open a PR?

**Comment by ehuaa** [comment]
> I think it comes down to just adding `sliding_window` to the call for `_prepare_4d_causal_attention_mask_for_sdpa` yes. Would you like to open a PR?

Sure，and i'll open a PR later in this week

**Comment by cyr0930** [comment]
any plan for pr?

**Comment by ArthurZucker** [comment]
#29407 should fix this issue

**Comment by cyr0930** [comment]
@ArthurZucker Oh you are right. Thanks.

**Comment by fxmarty** [comment]
Fixed in https://github.com/huggingface/transformers/pull/30127
